Add Pow, Ln, and Exp functions

sam-vdp opened the pull request:
Add Exp and Ln functions and Pow based on the combination of both. Ln uses Newton's method to approximate the solution, this code was ported from libfixmath, which your license already covers. The Exp function uses a power series

Commit:
Add Pow, Ln, and Exp functions

asik commented on src/Fix64.cs:
```diff
@@ -359,6 +368,94 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
+        {
+            if (x.RawValue == 0) return One;
+            if (x == One) return E;
+            if (x >= LnMax) return MaxValue;
```
What are the checks against LnMax and LnMin for? The tests still pass without. Optimization?

sam-vdp replied:
Yes. I figured if we know precisely where the argument range of the function ends, might as well test for it, but I guess if the early bailout test rarely succeeds it's a waste of cycles. I'm fine with removing it, your call.

asik replied:
It's a good optimization I think, just wanted to make sure.

asik reviewed (changes requested):
Thank you so much for this contribution! I have made some comments here.

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
+                if (b <= Fix64.Zero)
+                    continue;
+
+                if (b > Fix64.MaxValue / (Fix64)100)
```
Probably a leftover from a previous algorithm I tried that didn't handle large arguments well. Removed.

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
+                    continue;
+
+                // Reduced precision requirements for very small values
+                double maxDelta = b < (Fix64)0.000001m ? 0.1 : (double)(Fix64.Precision * 10);
```
The worst delta actually seen from the test data is 0.011, I've adjusted the test to reflect that. My guess would be that the errors of the argument scaling (repeated multiplication with E^4) add up and taint the result?

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
+            {
+                var b = Fix64.FromRaw(m_testCases[i]);
+
+                if (b < Fix64.Zero)
```
Done.

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
+
+                if (b < Fix64.Zero)
+                    continue;
+                if (b > Fix64.MaxValue / (Fix64)100)
```
For very large and very small bases the precision becomes bad or terrible, mostly because Ln isn't very precise anymore.
Example:
`Pow(0.0000000002328306436538696289, -0.5) = expected 65536 but got 65901.3926025061`
Or somewhat pathological cases like this where it really falls apart:
`Pow(2147483647.9999999997671693563, 0.9999999997671693563461303711) = expected 2147483637.25622 but got 1426713511.50211`

Finally, maybe the error should be measured relatively, instead of absolutely? 
`Pow(2147483647.9999999997671693563, 0.7911919844336807727813720703) = expected 24173977.2437025 but got 24173942.9738885`

I don't see a way to fully fix these cases, but also describing the exact limitations is somewhat difficult. I have refactored the test somewhat and am looking forward to your comments.

asik replied:
So, the test skips half the range [0..MaxValue], and in the range ]1.. MaxValue/2] the delta is artificially reduced to a fraction of itself. I'm not sure I understand the rationale behind this? In any case I'm not comfortable with an operation that's effectively untested over a large range of inputs and boundary conditions (even though I'm guilty of that myself e.g. Tan - I do document that it's not tested well though).

If Ln is the culprit then let's find a more accurate Ln. You got me reading on the topic :) I think an accurate base-2 logarithm for a 64-bit integer would be a good start. https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog has ideas, https://stackoverflow.com/questions/11376288/fast-computing-of-log2-for-64-bit-integers as well. 

Once we have that, Fix64 is just a long divided by 2^32, and log(x/y) is log(x) - log(y), so log2 of a Fix64 is log2(RawValue) - 32 (right?)

Once we have an accurate log2, ln x is just log2 x/log2 e.

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
+            {
+                var b = Fix64.FromRaw(m_testCases[j]);
+
+                if (b <= Fix64.Zero)
```
Done.

Commit:
Fix range check on Ln(0), refactor tests

sam-vdp commented on src/Fix64.cs:
```diff
@@ -359,6 +368,94 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
+            Fix64 result = x + One;
+            Fix64 term = x;
+
+            for (int i = 2; i < 40; i++)
```
WRT to your comment below: Yes, a better Ln implementation would be great! However, in your comment below, I think it's half of the solution. The code in the links calculates only the integer part of the log2, but we'll still need the fractional part, right?
Wikipedia has an algorithm for that, but I would assume comes with a similar speed and precision issues:
https://en.wikipedia.org/wiki/Binary_logarithm#Iterative_approximation

This one looks promising, I'll have a look at that:
https://github.com/dmoulding/log2fix

Commit:
Add Log2, Pow2 algorithms, base Pow on them

Commit:
Include log2fix in LICENSE.txt

sam-vdp commented:
Thank you very much for pushing back on the crappy previous solution and for your very good idea of using log2. I've added Log2 and Ln based on it, which don't have the previous precision issues.

Furthermore I've removed Exp and added Pow2, which splits the exponentiation into an integer and fractional part and handles large values much better, while maintaining precision for smaller values.

Please note that the Pow test still has some heuristics regarding the expected precision. Firstly, for large results the absolute errors become larger, which needs to be compensated. Secondly, some test cases are slightly pathological. E.g. `Pow(1-Fix64.FromRaw(1), Fix64.MaxValue)` is evaluated by doing `Pow2(Fix64.MaxValue * Fix64.Log2(1-Fix64.FromRaw(1))`, which becomes `Pow2(Fix64.MaxValue * Fix64.Precision)` which of course is not very precise.

asik commented on tests/Fix64Tests.cs:
```diff
@@ -428,12 +445,12 @@ public void Pow()
                     {
                         var expected = e == Fix64.Zero ? 1 : b == Fix64.Zero ? 0 : Math.Min(Math.Pow((double)b, (double)e), (double)Fix64.MaxValue);
```
This is hiding at least one bug: it ignores the case Pow(0, negative power) which should give MaxValue (the closest we have to infinity).

sam-vdp replied:
Mathematically, 0^x (x<0) is undefined, so I would argue 0, MaxValue or Infinity are pretty much equally wrong. Windows Calculator for example returns 0. I've added an exception for this case, after all, the division operator also throws instead of return MaxValue.

asik commented on tests/Fix64Tests.cs:
```diff
@@ -428,12 +445,12 @@ public void Pow()
-                            continue;
+                        // Absolute precision deteriorates with large result values, take this into account
+                        // Similarly, large exponents reduce precision, even if result is small.
+                        double maxDelta = Math.Abs((double)e) > 100000000 ? 0.5 : expected > 100000000 ? 10 : expected > 1000 ? 0.5 : 0.00001;
```
Actually if the "real" Pow(x, y) > MaxValue then you expect exactly MaxValue with 0 delta.

sam-vdp replied:
That would be nice, but actually the Pow approximation is still subject to the limitations that increase the maxDelta to 0.5 before. I.e. even if the expected value is MaxValue, the actual result may be slightly below, but inside the 0.5 delta.

asik commented on src/Fix64.cs:
```diff
@@ -382,66 +395,68 @@ public static Fix64 Exp(Fix64 x)
         }
 
-        public static Fix64 Ln(Fix64 x)
+        public static Fix64 Log2(Fix64 x)
```
I wouldn't expose Log2 as public API. A more useful API (which can be a separate PR) would be Log(a, newBase) like System.Math offers.

asik commented on src/Fix64.cs:
```diff
@@ -382,66 +395,68 @@ public static Fix64 Exp(Fix64 x)
+            return Fix64.FromRaw(y);
+        }
 
+        public static Fix64 Ln(Fix64 x)
```
All public APIs need a ///summary. I try to stay close to System.Math in descriptions. Also discussion of accuracy if it's not accurate down to Fix64.Precision (see for ex. Sin in the latest master - I just updated it yesterday).

sam-vdp commented on tests/Fix64Tests.cs:
```diff
@@ -387,18 +410,19 @@ public void Ln()
+        public void Pow2()
         {
-            double maxDelta = (double)(Fix64.Precision * 10);
+            var test = (double)Fix64.Pow2((Fix64)3);
```
Remove

sam-vdp commented on src/Fix64.cs:
```diff
@@ -99,6 +98,14 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
+        /// <summary>
+        /// Returns the fractional part of the specified number.
+        /// </summary>
+        public static Fix64 FractionalPart(Fix64 value)
```
Removed

sam-vdp commented on src/Fix64.cs:
```diff
@@ -368,12 +375,18 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
         }
 
-        public static Fix64 Exp(Fix64 x)
+        public static Fix64 Pow2(Fix64 x)
```
The thing with Pow2 and Log2 is, that I would still like to have dedicated tests for them, not just their combined results. With the tests in the separate assembly, I don't see how that would be possible without making them public?

asik replied:
Good point. Make it internal and add a InternalsVisibleTo attribute to the assembly in an AssemblyInfo.cs file.

sam-vdp replied:
That's a neat trick. Done.

Commit:
Removed public constants, remove FractionalPart API, use FastMul where appropriate, add documentation to public APIs, add DivideByZeroException to Pow(0, -x)

Commit:
Make Pow2 and Log2 internal

asik commented:
I've made a few stylistic changes locally and merged. Thank you for this contribution!
